activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb

```diff
@@ -29,5 +27,42 @@ def synchronize(&block)
         end
       end
     end
+    # A monitor that will permit dependency loading while blocked waiting for
+    # the lock.
+    class LoadInterlockAwareMonitor < Monitor
+      include LoadInterlockAwareMonitorMixin
+    end
+
+    class ThreadLoadInterlockAwareMonitor # :nodoc:
+      prepend LoadInterlockAwareMonitorMixin
+
+      def initialize
+        @owner = nil
+        @count = 0
+        @mutex = Mutex.new
+      end
+
+      private
+        def mon_try_enter
+          if @owner != Thread.current
+            return false unless @mutex.try_lock
+            @owner = Thread.current
+          end
+          @count += 1
+        end
+
+        def mon_enter
+          @mutex.lock if @owner != Thread.current
+          @owner = Thread.current
+          @count += 1
+        end
+
+        def mon_exit
```

Comment:
I think this will quietly go negative if misused... which is low probability given our intended usage, but still maybe worth a touch more protection. Maybe just `raise if @owner != Thread.current`?